File: experiments_realworld/recursion/problem03/implementations/optimal/solution.cpp
#include <bits/stdc++.h>
using namespace std;

// RECURSIVE DFS rerooting (deep recursion), C++ counterpart of the Python version.
// For each node, the MAXIMUM distance to any other node. Two recursive passes:
//   dfs_down: down1[u] = longest downward path, down2[u] = second longest
//             downward path (through a DIFFERENT child); arg1[u] = the child
//             giving down1 (to exclude it for that child in the up pass).
//   dfs_up:   up[v] = 1 + max(up[u], best downward of u excluding branch v).
//   ans[u] = max(down1[u], up[u]).
// Same algorithm as the Python version; only the language differs.

int n;
vector<vector<int>> adj;
vector<int> down1, down2, arg1, up_;

void dfs_down(int u, int parent) {
    for (int v : adj[u]) {
        if (v != parent) {
            dfs_down(v, u);
            int cand = down1[v] + 1;
            if (cand > down1[u]) {
                down2[u] = down1[u];
                down1[u] = cand;
                arg1[u] = v;
            } else if (cand > down2[u]) {
                down2[u] = cand;
            }
        }
    }
}

void dfs_up(int u, int parent) {
    for (int v : adj[u]) {
        if (v != parent) {
            int best_excl = (arg1[u] == v) ? down2[u] : down1[u];
            up_[v] = max(up_[u], best_excl) + 1;
            dfs_up(v, u);
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n;
    adj.assign(n + 1, {});
    for (int e = 0; e < n - 1; e++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    if (n == 1) {
        cout << 0 << "\n";
        return 0;
    }

    down1.assign(n + 1, 0);
    down2.assign(n + 1, 0);
    arg1.assign(n + 1, 0);
    up_.assign(n + 1, 0);

    dfs_down(1, 0);
    dfs_up(1, 0);

    for (int i = 1; i <= n; i++) {
        cout << max(down1[i], up_[i]);
        if (i < n) cout << " ";
    }
    cout << "\n";

    return 0;
}

File: experiments_realworld/recursion/problem03/implementations/optimal/solution.py
import sys
sys.setrecursionlimit(300000)

# RECURSIVE DFS rerooting (deep recursion). For each node, the MAXIMUM distance
# to any other node. Two recursive passes:
#   dfs_down: for each u, down1[u] = longest downward path, down2[u] = second
#             longest downward path (through a DIFFERENT child) -- needed for the
#             reroot.
#   dfs_up:   up[u] = longest path going through the parent; for a child c it is
#             1 + max(up[u], down-of-u-excluding-c). ans[u] = max(down1[u], up[u]).
# Genuine recursive tree rerooting; recursion depth up to n = 2*10^5 (chain).

def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    adj = [[] for _ in range(n + 1)]
    idx = 1
    for _ in range(n - 1):
        a = int(data[idx]); b = int(data[idx + 1]); idx += 2
        adj[a].append(b)
        adj[b].append(a)

    if n == 1:
        print(0)
        return

    down1 = [0] * (n + 1)   # longest downward path from u
    down2 = [0] * (n + 1)   # second longest downward path (different child)
    arg1 = [0] * (n + 1)    # child giving down1 (to exclude it for that child)
    up = [0] * (n + 1)      # longest path going upward through the parent

    def dfs_down(u, parent):
        for v in adj[u]:
            if v != parent:
                dfs_down(v, u)
                cand = down1[v] + 1
                if cand > down1[u]:
                    down2[u] = down1[u]
                    down1[u] = cand
                    arg1[u] = v
                elif cand > down2[u]:
                    down2[u] = cand

    def dfs_up(u, parent):
        for v in adj[u]:
            if v != parent:
                # best downward of u excluding the branch through v
                best_excl = down2[u] if arg1[u] == v else down1[u]
                up[v] = max(up[u], best_excl) + 1
                dfs_up(v, u)

    dfs_down(1, 0)
    dfs_up(1, 0)

    out = []
    for i in range(1, n + 1):
        out.append(str(max(down1[i], up[i])))
    sys.stdout.write(" ".join(out) + "\n")

if __name__ == "__main__":
    main()
